tests/fake/test_fake_upload.py

```diff
@@ -0,0 +1,78 @@
+import sys
+
+import pytest
+
+from pubtools.pulplib import FakeController, FileRepository, PulpException
+
+
+def test_can_upload(tmpdir):
+    """repo.upload_file() succeeds with fake client and populates upload_history."""
+    controller = FakeController()
+
+    controller.insert_repository(FileRepository(id="repo1"))
+
+    client = controller.client
+    repo1 = client.get_repository("repo1").result()
+
+    somefile = tmpdir.join("some-file.txt")
+    somefile.write(b"there is some binary data:\x00\x01\x02")
+
+    upload_f = repo1.upload_file(str(somefile))
+
+    # The future should resolve successfully
+    tasks = upload_f.result()
+
+    # The task should be successful.
+    assert tasks[0].succeeded
+
+    # The change should be reflected in the controller's upload history
+    history = controller.upload_history
+
+    digest = "fad3fc1e6d583b2003ec0a5273702ed8fcc2504271c87c40d9176467ebe218cb"
+    assert len(history) == 1
+    assert history[0].repository == repo1
+    assert history[0].tasks == tasks
+    assert history[0].name == str(somefile)
```

Comment:
Shouldn't it actually be just "some-file.txt" - I mean just the basename and not the entire path with leading tmpdir?

```suggestion
    assert history[0].name == "some-file.txt"
```

If tests don't pass with that change then something might be wrong.

Reply:
Hm, I feel this should be the full path. 
The 'name' here is actually unit_key["name"], which is relative_url passed to import, which will be used as the remote path, so I think the full path should be used. 
And the behavior in fake client should be consistent.

Reply:
OK, we might not have had the same idea over how the API should actually work then.

IMHO, in the case where the caller didn't provide relative_url during upload, the full path to the file locally should NOT be used to set relative_url, only the basename should be used.  That's because the full path is almost certain to be meaningless.

For example if the user prepared some content to push under: /mnt/staging/my-push-123/repo1/ISO/rhel-boot-dvd.iso

Then, while it could make sense to default the filename to "rhel-boot-dvd.iso" during upload, I don't think it could make sense to include that whole /mnt/staging/... path.

In practice, relative_url 99% of the time would be only a basename I think.